File: lib/to_mdx/parsers/open_document/text/span.rb
# frozen_string_literal: true

require_relative "../base"

module ToMdx
  module Parsers
    module Text
      class Span < Base
        attr_reader :style_name

        def initialize(element = nil, context = {})
          super(element, context)
          @style_name = context[:styles]&.use(element&.attributes&.[]("text:style-name")) || ""
        end

        def to_mdx
          content = contentful_children.map(&:to_mdx).join
          if @style_name.empty?
            content
          else
            "<span className=\"#{@style_name}\">#{content}</span>"
          end
        end

        def to_s
          content = contentful_children.map(&:to_s).join
          if context[:parent_is_code_snippet] && is_plaintext? && has_background?
            "___#{content}___"
          else
            content
          end
        end

        def is_plaintext?
          single_child? { |child| child.is_a?(Parsers::Text::PlainText) }
        end

        def has_background?
          return false unless @context[:styles]
          style_properties = @context[:styles].properties(element&.attributes&.[]("text:style-name"))
          !!style_properties["background-color"]
        end

        register_for "text:span"
      end
    end
  end
end

File: src/parsers/open-document/text/span.ts
import type { Style } from "../../../styles.ts";
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";
import { PlainText } from "./plain_text.ts";

export class Span extends BaseElement {
  readonly styleName: Style;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.styleName = context.styles.use(this.attr("text:style-name"));
  }

  toMdx(): string {
    const content = this.contentfulChildren()
      .map((child) => child.toMdx())
      .join("");

    if (this.styleName.empty()) {
      return content;
    }

    const tag = this.context("isSvg") ? "tspan" : "span";
    return `<${tag} className="${this.styleName}">${content}</${tag}>`;
  }

  toString(): string {
    const content = this.contentfulChildren()
      .map((child) => child.toString())
      .join("");

    if (this.isPlaintext() && this.hasBackground()) {
      return `___${content}___`;
    } else {
      return content;
    }
  }

  isPlaintext(): boolean {
    return this.singleChild((child) => child instanceof PlainText);
  }

  private hasBackground(): boolean {
    // We get the attribute from the element, not the style name member, because we don't want to process it yet
    const styleName = this.attr("text:style-name");
    if (!styleName) {
      return false;
    }

    const styleProperties = this.context("styles").get(styleName);
    return !!styleProperties["background-color"];
  }

  static {
    BaseElement.registerFor.call(this, "text:span");
  }
}
